Add `unique!`

JackDevine opened the pull request:
Added a `unique!` function that creates a fast path for sorted data.
Also added some tests for the function.

Continuation of #20576. Fixes https://github.com/JuliaLang/julia/issues/20549.

JackDevine committed:
Added `unique!` #20576

Added a `unique!` function that creates a fast path for sorted data.
Also added some tests for the function.

JackDevine committed:
Added `unique!` #20576

Added a `unique!` function that creates a fast path for sorted data.
Also added some tests for the function.

mbauman commented on base/set.jl:
```diff
@@ -191,6 +191,62 @@ function unique(f::Callable, C)
+        # When we loop through c, we only need to remember one state and one
+        # item in the collection.
+        currentstate = start(idxs)
+        currentitem = c[currentstate]
```
This will fail for `unique!([])`.  Manually using the iteration protocol is a little tedious, but you always need to call `done(idxs, currentstate)` before getting the next element with `idx, currentstate = next(idxs, currentstate)`. And then you can't really use `currentstate` for anything beyond those two functions (`done` and `next`).  It's an internal implementation detail… and you can't count on it being useful as an index itself.

JackDevine replied:
Right, originally I had a function more like this:
```julia
function uniquejd!(itr)  # I used my initials to distinguish the function.
    seen = Set{eltype(itr)}()
    count = 0  # Number of unique values seen so far.
    for x in itr
        @inbounds if ∉(x, seen)
            count += 1            
            push!(seen, x)
            itr[count] = x
        end
    end
    resize!(itr, count)
end
```
Shall I revert to that idea? It is also really easy to deal with the sorted case using the above idea.

JackDevine replied:
Right, I used to have something like this:

```julia
function unique!(itr)
    seen = Set{eltype(itr)}()
    count = 0  # Number of unique values seen so far.
    for x in itr
        @inbounds if ∉(x, seen)
            count += 1            
            push!(seen, x)
            itr[count] = x
        end
    end
    resize!(itr, count)
end
```
Should I revert back to that version? I can make that work with the fast path too.

JackDevine replied:
Or we could just add something like
```julia
if isempty(c)
    return c
end
```
At the start.

StefanKarpinski replied:
If you limit this to `AbstractVector` then you can use `for i in indices(v,1)` to iterate from the first index to the last. I would argue that `AbstractVector` is the only built-in data structure that it makes sense to define this for. For ordered collections the "keys" are integers, in which case the collection is a kind of `AbstractVector`. The only other built-in general collection types are `Dict` and `Set` and `unique!` doesn't make sense for those.

mbauman replied:
I think the idea here is good... you just have to add the calls to done and use `idx` as the index instead of `currentstate`.

JackDevine replied:
Thanks for all of your help, from what I understand from above, this is what I managed to hack together
```julia
function unique!(c)
    if issorted(c) || issorted(c, rev=true)
        done(c, 1) ? count = 0 : count = 1
        for i in indices(c, 1)
            x = c[i]
            currentitem = c[count]
            if x != currentitem
                count += 1
                c[count] = x
            end
        end
    else
        seen = Set{eltype(c)}()
        idxs = eachindex(c)
        m = n = start(idxs)
        count = 0
        while !done(c, n)
            i, n = next(idxs, n)
            x = c[i]
            if x ∉ seen
                count += 1
                push!(seen, x)
                j, m = next(idxs, m)
                c[j] = x
            end
        end
    end
    resize!(c, count)
end
```

StefanKarpinski commented on base/set.jl:
````diff
@@ -191,6 +191,62 @@ function unique(f::Callable, C)
+```
+"""
+function unique!(c)
+    if issorted(c) || issorted(reverse(c))
````
Reversing the collection is expensive. Instead you should do `issorted(c, rev=true)`.

JackDevine committed:
Fixed `unique!` (#20576 #20619)

returns a zero element array for []

JackDevine committed:
Fixed `unique!` (#20576 #20619)

returns a zero element array for []

JackDevine committed:
Limited the type to `AbstractVector`

Sorry that I left this for so long. The current implementation is
limited to `AbstractVector` and has the changes that @mbauman
requested. I also have some bechmarks to show the performance of
`unique!` vs `unique`

```julia
using BenchmarkTools

A = [rand(1:10) for _ in 1:100_000]
B = [rand(1:1000) for _ in 1:100_000]

println("Test performance on a large vector with not many unique
elements.")
println("Non mutating unique():")
@btime unique(A)
println("Mutating unique()\n")
@btime unique!(A)

println("Test performance on a large vector with quite a few unique
elements.")
println("Non mutating unique():")
@btime unique(B)
println("Mutating unique():")
@btime unique!(B)

Test performance on a large vector with not many unique elements.
Non mutating unique():
  1.456 ms (9 allocations: 848 bytes)
Mutating unique():
  1.750 μs (6 allocations: 592 bytes)

Test performance on a large vector with quite a few unique elements.
Non mutating unique():
  1.477 ms (28 allocations: 66.06 KiB)
Mutating unique():
  56.178 μs (19 allocations: 49.70 KiB)
```

JackDevine committed:
Limited the type to `AbstractVector`

Sorry that I left this for so long. The current implementation is
limited to `AbstractVector` and has the changes that @mbauman
requested. I also have some bechmarks to show the performance of
`unique!` vs `unique`

```julia
using BenchmarkTools

A = [rand(1:10) for _ in 1:100_000]
B = [rand(1:1000) for _ in 1:100_000]

println("Test performance on a large vector with not many unique
elements.")
println("Non mutating unique():")
@btime unique(A)
println("Mutating unique()\n")
@btime unique!(A)

println("Test performance on a large vector with quite a few unique
elements.")
println("Non mutating unique():")
@btime unique(B)
println("Mutating unique():")
@btime unique!(B)

Test performance on a large vector with not many unique elements.
Non mutating unique():
  1.456 ms (9 allocations: 848 bytes)
Mutating unique():
  1.750 μs (6 allocations: 592 bytes)

Test performance on a large vector with quite a few unique elements.
Non mutating unique():
  1.477 ms (28 allocations: 66.06 KiB)
Mutating unique():
  56.178 μs (19 allocations: 49.70 KiB)
```

JackDevine commented:
@StefanKarpinski @mbauman , I am not sure why travis didn't pass. Would it be easier if I just deleted this branch and started again?

nalimilan commented:
No, keeping this PR is fine. I've restarted the build (it had failed due to a timeout).

nalimilan commented on base/set.jl:
```diff
@@ -191,6 +191,61 @@ function unique(f::Callable, C)
 
 """
+    unique!(c)
+Remove duplicate items and return the modified collection, as determined by
```
Missing blank line.

JackDevine replied:
Thanks!

JackDevine committed:
Added a blank line to the docstring.

JackDevine committed:
Added a blank line to the docstring.

JackDevine commented:
Hmm, Travis is still not passing, any ideas why?

nalimilan commented:
Hmm, looks like another unrelated error:
```
From worker 1:		From worker 1:	Worker 5 failed running test spawn:

Some tests did not pass: 103 passed, 0 failed, 1 errored, 4 broken.spawn: Error During Test

  Got an exception of type LoadError outside of a @test

  LoadError: InterruptException:

  while loading /private/tmp/julia/share/julia/test/spawn.jl, in expression starting on line 425

```
https://gist.github.com/nalimilan/6783ae43c596828975ebd0c26f5e9b73

I've restarted the build again.

JackDevine commented:
Thanks @nalimilan !

nalimilan commented on base/set.jl:
```diff
@@ -191,6 +191,62 @@ function unique(f::Callable, C)
+        # If c is sorted, then we only need to keep track of one element and add that to c
+        # everytime that we see a new element.
+        count = 1
+        for i in indices(c, 1)
```
Should probably use `@inbounds` here? Maybe also in the other branch, not sure.

Also, `indices(c, 1)` could just be `eachindex` since `c` is a vector.

JackDevine replied:
I just benchmarked the `@inbounds` idea and it didn't make any difference on my machine, but all of the tests still pass. Using `indices` was Stefan's idea, but I think that we should change it to `eachindex`, because that does seem more simple.

nalimilan commented on base/set.jl:
````diff
@@ -191,6 +191,62 @@ function unique(f::Callable, C)
+```
+"""
+function unique!(c::AbstractVector)
+    if (issorted(c) || issorted(c, rev=true)) && !isempty(c)
````
`isempty(c)` could be a separate first no-op branch to avoid all the unnecessary operations.

JackDevine replied:
Thanks!

nalimilan commented on base/set.jl:
```diff
@@ -191,6 +191,62 @@ function unique(f::Callable, C)
 end
 
 """
+    unique!(c)
```
Should add `::AbstractVector` since that's the only method for now. Then the description below shouldn't say "collection" : just say `c`. Also, it would be clearer if "as determined..." was right after "duplicate items" (like in `?unique`).

JackDevine replied:
Should we also change the name from c to something like itr or A?

nalimilan replied:
I guess `A` or `a` would make more sense as long as this only applies to vectors.

JackDevine committed:
Created a no-op branch for an empty array and updated docstring.

JackDevine committed:
Created a no-op branch for an empty array and updated docstring.

nalimilan commented on base/set.jl:
```diff
@@ -191,6 +191,63 @@ function unique(f::Callable, C)
+"""
+function unique!(c::AbstractVector)
+    if isempty(c)
+        count = 0
```
You can just do `return c`, which is clearer and slightly faster.

JackDevine replied:
Good point, fixed.

nalimilan commented on base/set.jl:
```diff
@@ -191,6 +191,63 @@ function unique(f::Callable, C)
+function unique!(c::AbstractVector)
+    if isempty(c)
+        count = 0
+    elseif (issorted(c) || issorted(c, rev=true))
```
Parentheses are no longer needed.

JackDevine committed:
Changed the name from c to A and tidied up the code.

JackDevine committed:
Changed the name from c to A and tidied up the code.

JackDevine commented:
Thanks for all of your help people, I have rerun the benchmarks on my windows machine, which has the following specs:
64 bit OS,
Intel i7-4790K CPU @4.00GHz,
16.0GB ram

```julia
using BenchmarkTools

srand(42)
A = [rand(1:10) for _ in 1:100_000]
srand(42)
B = [rand(1:1000) for _ in 1:100_000]

println("Test performance on a large vector with not many unique elements.")
println("Non mutating unique():")
@btime unique(A)
println("Mutating unique()\n")
@btime unique!(A)

println("Test performance on a large vector with quite a few unique elements.")
println("Non mutating unique():")
@btime unique(B)
println("Mutating unique():")
@btime unique!(B)

Test performance on a large vector with not many unique elements.
Non mutating unique():
  1.046 ms (14 allocations: 1.19 KiB)
Mutating unique()
  1.873 μs (11 allocations: 960 bytes)
Test performance on a large vector with quite a few unique elements.
Non mutating unique():
  1.063 ms (59 allocations: 67.95 KiB)
Mutating unique():
  48.239 μs (50 allocations: 51.59 KiB)
```
So we have a factor 1000 speed up as well as a reduction in allocations.

StefanKarpinski commented:
Excellent :D

JackDevine committed:
Changed to `eachindex ` and made use of the `@inbouds` macro.

JackDevine committed:
Changed to `eachindex ` and made use of the `@inbouds` macro.

tkelman commented on base/set.jl:
```diff
@@ -191,6 +191,63 @@ function unique(f::Callable, C)
 end
 
 """
+    unique!(A::AbstractVector)
```
this needs to be added to the stdlib doc index to show up in the rendered html docs

fredrikekre replied:
(doc/src/stdlib/collections.md line 82)

JackDevine replied:
Done.

tkelman commented on base/set.jl:
```diff
@@ -191,6 +191,63 @@ function unique(f::Callable, C)
+        return A
+    elseif issorted(A) || issorted(A, rev=true)
+        # If A is sorted, then we only need to keep track of one element and add that to A
+        # everytime that we see a new element.
```
2 words, "every time"

JackDevine replied:
Doh!

JackDevine committed:
Made every time two words.

JackDevine committed:
Made every time two words.

JackDevine committed:
Added `unique!` to stdlib doc index

JackDevine committed:
Added `unique!` to stdlib doc index

JackDevine commented:
Hmm, Travis is passing, but now appveyor is failing. I looked at the log and among other things it said:
```
if ($env:APPVEYOR_PULL_REQUEST_NUMBER -and $env:APPVEYOR_BUILD_NUMBER -ne ((Invoke-RestMethod ` https://ci.appveyor.com/api/projects/$env:APPVEYOR_ACCOUNT_NAME/$env:APPVEYOR_PROJECT_SLUG/history?recordsNumber=50).builds | ` Where-Object pullRequestId -eq $env:APPVEYOR_PULL_REQUEST_NUMBER)[0].buildNumber) { ` throw "There are newer queued builds for this pull request, failing early." }
```
Is it complaining that I pushed too many commits too quickly?

JackDevine commented:
Also, if I test the performance of the `@inbounds`, then the best I can get is the following:

```julia
srand(42)
B = [rand() for _ in 1:1_000_000]
@btime unique!(B)
```
With `@inbounds` I get
```
107.309 ms (122 allocations: 34.51 MiB)
```
and without it I get
```
115.333 ms (122 allocations: 34.51 MiB)
```
Which is not a great improvement, things don't get much better if I use `@inbounds begin ... end` to wrap both of the loops. Maybe we should consider not using `@inbounds`?

nalimilan commented on doc/src/stdlib/collections.md:
```diff
@@ -79,6 +79,7 @@ Base.eltype
 Base.indexin
 Base.findin
 Base.unique
+Base.unique!(::AbstractVector)
```
I don't think `::AbstractVector` is needed here. If more methods are added later, they would have to be included here too, so better drop it.

JackDevine replied:
Good point, fixed.

nalimilan commented on test/sets.jl:
```diff
@@ -216,6 +216,17 @@ u = unique([1,1,2])
+    unique!(u)
+    @test u == [1,3,2]
+    @test unique!([]) == []
+    u = [1,2,2,3,5,5]
```
Could also be a good idea to test a vector sorted in decreasing order?

JackDevine replied:
Agreed.

nalimilan commented on base/set.jl:
```diff
@@ -191,6 +191,63 @@ function unique(f::Callable, C)
+    elseif issorted(A) || issorted(A, rev=true)
+        # If A is sorted, then we only need to keep track of one element and add that to A
+        # every time that we see a new element.
+        count = 1
```
To support arrays whose indices do not start at `1`, better use the same strategy as for the other branch via `m = start(idxs)`, etc. Then you can compute `count` at the end as `m-start(idxs)`.

Also, putting `@inbounds` before the `for` will enable it for the whole block, including `A[count]`.

JackDevine replied:
Shouldn't it be `m-start(idxs)-1`? On this topic should we make a test for indices not starting at one? I can think of some, bu they would require us to overload `Base.getindex` etc and I am not too sure of that is standard practice.

tkelman replied:
there's an offset array type defined in TestHelpers that gets used in a few places in the tests

nalimilan replied:
You meant `+1`? Yeah, I guess, I wasn't sure what the exact code would look like. As for testing arrays with indices not starting at 1, maybe @timholy can help.

JackDevine replied:
Yes, I meant +1, sorry. I will get started on making an OffSetArray test now.

JackDevine replied:
@tkelman ,  `resize!` has no method for `OffSetArray`, so I can't get `unique!` to work on it.

tkelman replied:
maybe you could add one?

JackDevine replied:
I have made a separate PR for `resize!` and `done`.

nalimilan commented:
> Is it complaining that I pushed too many commits too quickly?

That's not an issue, it just means that a build was skipped because a new commit was pushed before it started.

Regarding `@inbounds`, better keep it even if the difference is small, it doesn't hurt. Also, you can put it right before `for`, which will enable it in more places than currently (in both branches).

tkelman commented:
> Regarding `@inbounds`, better keep it even if the difference is small, it doesn't hurt

It does hurt if a user-defined type hasn't implemented its `eachindex` / `getindex` combination correctly and gets segfaults instead of a bounds error.

JackDevine committed:
Implemented suggestions

Removed `@inbounds`, allowed support for vectors with indices not
starting at one and added a test for arrays sorted in descending order.

JackDevine committed:
Implemented suggestions

Removed `@inbounds`, allowed support for vectors with indices not
starting at one and added a test for arrays sorted in descending order.

nalimilan commented:
> It does hurt if a user-defined type hasn't implemented its eachindex / getindex combination correctly and gets segfaults instead of a bounds error.

Yeah, but an incorrectly implemented custom type will get bugs in many other places anyway. From https://github.com/JuliaLang/julia/issues/20469 I understood that the policy is that `AbstractArray` implementations which use the `@boundscheck` annotation are responsible for getting it right.

JackDevine committed:
Removed trailing whitespace.

JackDevine committed:
Removed trailing whitespace.

nalimilan commented on base/set.jl:
```diff
@@ -191,6 +191,65 @@ function unique(f::Callable, C)
+        for i in idxs
+            x = A[i]
+            if x != A[m]
+                _, m = next(idxs, m)
```
In general doing this isn't correct: `m` is the state of the iterator, which can be different from the index. So you need to store both the index and state in separate variables, as you do in the other branch. Hopefully in most cases the compiler will be able to elide redundant computations.

JackDevine replied:
Ok, so I should write `j, m = next(idxs, m)` even though `j` is not being used?

nalimilan replied:
You have to use `j` to index into `A`. `m` is not a valid index, it's the state of the iterator, which could be anything.

JackDevine replied:
Right, sorry that was my bad. I have sort of fixed this but there is one point where I have to use `j += 1`, so that will only work if the indices are in ascending order.

JackDevine committed:
Removed trailing whitespace and keep track of indeces.

JackDevine committed:
Removed trailing whitespace and keep track of indeces.

JackDevine committed:
Use the index rather than the state to index `A`.

JackDevine committed:
Use the index rather than the state to index `A`.

JackDevine committed:
Made `unique!` compatible with OffsetArrays.

JackDevine committed:
Made `unique!` compatible with OffsetArrays.

JackDevine committed:
Revert "Made `unique!` compatible with OffsetArrays."

This reverts commit 763ef1888a28c8eaea5ace7bde1eba3ff0456101.

JackDevine committed:
Revert "Made `unique!` compatible with OffsetArrays."

This reverts commit 763ef1888a28c8eaea5ace7bde1eba3ff0456101.

JackDevine committed:
Made `unique!` compatible with OffsetArrays but didnlt make the OffsetArrays tests.

JackDevine committed:
Made `unique!` compatible with OffsetArrays but didnlt make the OffsetArrays tests.

JackDevine commented:
Just to prove that the OffsetArray stuff is working, if you have the code in TestHelpers and you include
```julia
Base.resize!(A::OffsetVector, nl::Integer) = resize!(A.parent, nl)
Base.done(A::OffsetVector, ind::Integer) = done(A.parent, ind - A.offsets[1])
```
Then the following tests work:
```julia
u = OffsetArray([1,2,2,3,5,5], -1)
unique!(u)
@assert u == OffsetArray([1,2,3,5], -1)

u = OffsetArray([5,5,4,4,2,2,0,-1,-1], -1)
unique!(u)
@assert u == OffsetArray([5,4,2,0,-1], -1)

u = OffsetArray([1,1,2,3,1,4,6], -1)
unique!(u)
@assert u == OffsetArray([1,2,3,4,6], -1)

u = OffsetArray([1,1,3,1,4,6], -1)
unique!(u)
@assert u == OffsetArray([1,3,4,6], -1)

u = OffsetArray(["w","we","w",5,"r",5,5], -1)
unique!(u)
@assert u == OffsetArray(["w","we",5,"r"], -1)
```

Something that I found when trying to make OffsetArrays work:
```julia
julia> unique([1,"w",1,3])
 3-element Array{Any,1}:
 1   
  "w"
 3

julia> unique!([1,"w",1,3])
 MethodError: no method matching isless(::String, ::Int64)
 Closest candidates are:
  isless(::AbstractFloat, ::Real) at operators.jl:98
  isless(::Real, ::Real) at operators.jl:266
  isless(::AbstractString, ::AbstractString) at strings/basic.jl:124

Stacktrace:
 [1] issorted(::Array{Any,1}, ::Base.Order.ForwardOrdering) at .\sort.jl:54
 [2] unique!(::Array{Any,1}) at .\In[3]:4
```
So checking that the array is sorted is causing problems. Although maybe this is not such a big deal because if you are using Any arrays then you probably aren't interested in the possible performance gains from `unique!`.

nalimilan commented:
Great! Just add the tests and the functions here, if they are identical `git` will rebase automatically once the other PR has been merged.

Though it's problematic to throw an error when some entries cannot be compared. I guess the only solution is to wrap the `issorted` calls inside `try... catch`. (We could also play with `method_exists`, but for that wouldn't work for `Any`, and even if a function exists it could throw an error.)

nalimilan commented on base/set.jl:
```diff
@@ -191,6 +191,66 @@ function unique(f::Callable, C)
+        # If A is sorted, then we only need to keep track of one element and add that to A
+        # every time that we see a new element.
+        idxs = eachindex(A)
+        m = first(A)
```
Better use another letter like `y` for consistency with `x`, and since `m` and `n` are used below to represent indices (which is consistent with standard math notation for integers).

nalimilan commented on base/set.jl:
```diff
@@ -191,6 +191,66 @@ function unique(f::Callable, C)
+        for i in idxs
+            x = A[i]
+            if x != m
+                m, j = next(idxs, j)
```
The value of `m` isn't actually used here, so just use `_` for clarity (and possibly efficiency).

nalimilan replied:
Woops, actually `m` and  `j` should be swapped here. I would do `j, state = next(idxs, state)`. `m` (to be renamed to `y`) shouldn't enter the picture.

nalimilan commented on base/set.jl:
```diff
@@ -191,6 +191,66 @@ function unique(f::Callable, C)
+        j = start(idxs)
+        for i in idxs
+            x = A[i]
+            if x != m
```
Should use `isequal` here to match the behavior of the other branch and the docs. Would be good to add tests with (repeated) `NaN`, `0.0` and `-0.0`, for both branches (i.e. tests which fail with the current code).

nalimilan commented on base/set.jl:
```diff
@@ -191,6 +191,66 @@ function unique(f::Callable, C)
+        idxs = eachindex(A)
+        m = n = start(idxs)
+        count = 0
+        while !done(A, n)
```
Come to think of it, this can be simplified to `for x in A` since `i` isn't needed. Then `j` can be renamed to `i`, and `m` would be clearer as `state`. You can also compute `count` after the loop as in the other branch to save one instruction per iteration.

JackDevine committed:
Fixed up `unique!` and added tests for OffsetArray

1. Added a `try ... catch` block to wrap the `issorted` calls, also
added a test for this case that would have failed under the previous
implementation.

2. More consistent naming of variables, `m` and `n` are now integers and
`x`, `y` and `item` are for states.

3. Replaced `!=` with `!isequal` to be more consistent with the docs
(and with `unique`). Added tests for this behaviour that would have
failed under the old implementation.

JackDevine committed:
Fixed up `unique!` and added tests for OffsetArray

1. Added a `try ... catch` block to wrap the `issorted` calls, also
added a test for this case that would have failed under the previous
implementation.

2. More consistent naming of variables, `m` and `n` are now integers and
`x`, `y` and `item` are for states.

3. Replaced `!=` with `!isequal` to be more consistent with the docs
(and with `unique`). Added tests for this behaviour that would have
failed under the old implementation.

JackDevine commented:
Thanks for the tips @nalimilan , at the moment the `try ... catch` block is in a little bit of an awkward place. I was thinking that maybe it should be after the `isempty` call, but I wanted your opinion first.

nalimilan commented:
> Thanks for the tips @nalimilan , at the moment the try ... catch block is in a little bit of an awkward place. I was thinking that maybe it should be after the isempty call, but I wanted your opinion first.

Looks OK to me, but maybe you should just do `isempty(A) && return A` at the top of the function. Also I would rename `Aissorted` to something shorter like `sorted` since there's no ambiguity.

nalimilan commented on base/set.jl:
```diff
@@ -191,6 +191,69 @@ function unique(f::Callable, C)
+        for i in idxs
+            x = A[i]
+            if !isequal(x, y)
+                _, j = next(idxs, j)
```
This is still incorrect: you need `j, state = next(idxs, state)`, and changing `j = start(idxs)` to `state = start(idxs)`.

nalimilan commented on base/set.jl:
```diff
@@ -191,6 +191,69 @@ function unique(f::Callable, C)
+                y = A[j] = x
+            end
+        end
+        count = j - start(idxs) + 1
```
`start(idxs)` should be `first(idx)` (same below).

nalimilan commented on base/set.jl:
```diff
@@ -191,6 +191,69 @@ function unique(f::Callable, C)
+        idxs = eachindex(A)
+        state = n = start(idxs)
+        for x in A
+            _, n = next(idxs, n)
```
This line has no effect, so just remove it and the one above defining `n`.

JackDevine committed:
Tidied up the try catch block and used the state of iterators correctly.

JackDevine committed:
Tidied up the try catch block and used the state of iterators correctly.

nalimilan commented on base/set.jl:
```diff
@@ -191,6 +191,69 @@ function unique(f::Callable, C)
+            x = A[i]
+            if !isequal(x, y)
+                j, state = next(idxs, state)
+                y = A[j+1] = x
```
It doesn't sound right to do `+1` here. Wouldn't it be better to add `_, state = next(idxs, state)` right before the loop instead?

JackDevine replied:
I agree, I just made the change and it means that `count = j - first(idxs) + 1` just like in the other branch.

JackDevine committed:
Added `resize!` to offset arrays and avoided using `j+1` inside the loop.

JackDevine committed:
Added `resize!` to offset arrays and avoided using `j+1` inside the loop.

nalimilan commented on test/TestHelpers.jl:
```diff
@@ -143,6 +143,7 @@ _offset(out, ::Tuple{}, ::Tuple{}) = out
 indsoffset(r::Range) = first(r) - 1
 indsoffset(i::Integer) = 0
 
+Base.resize!(A::OffsetVector, nl::Integer) = (resize!(A.parent, nl); A)
```
Add a line break since there was one before `end`.

nalimilan reviewed (dismissed):
Thanks, looks good to me!

Though I still think adding `@inbounds` would make sense. What do others think about it?

JackDevine committed:
Added a line break after the `resize` method.

JackDevine committed:
Added a line break after the `resize` method.

JackDevine commented:
With regards to `@inbounds`, I did some benchmarks with the latest version of `unique!`. One version didn't have `@inbounds` and the other had `@inbounds begin ... end` wrapping both loops.
```julia
using BenchmarkTools
srand(42)
A = [rand(1:100_000) for _ in 1:1_000_000] |> sort
srand(42)
B = [rand(1:100_000) for _ in 1:1_000_000]

println("Test the sorted branch.")
@btime unique!(A)

@assert A == unique(A)

println("Test the not sorted branch.")
@btime unique!(B)

@assert B == unique(B)
```
With `@inbounds`
```
Test the sorted branch.
  5.915 ms (298442 allocations: 4.55 MiB)
Test the not sorted branch.
  7.158 ms (198995 allocations: 6.04 MiB)
```
Without `@inbounds`
```
Test the sorted branch.
  6.426 ms (298442 allocations: 4.55 MiB)
Test the not sorted branch.
  7.324 ms (198995 allocations: 6.04 MiB)
```

JackDevine committed:
Made sure that variables remain inscope after the loops exited.

JackDevine committed:
Made sure that variables remain inscope after the loops exited.

tkelman commented:
instead of a try-catch, this could maybe use a dispatch approach similar to https://github.com/JuliaLang/julia/pull/22010

JackDevine committed:
Replaced the try-catch block with a dispatch approach.

JackDevine committed:
Replaced the try-catch block with a dispatch approach.

nalimilan commented:
Makes sense, but shouldn't we also apply the sorted method to `AbstractArray{<:Union{AbstractString, Symbol}}`?

Then there are other sortable types, like `Nullable` wrappers. Also, it would be really unfortunate that non-Base types would not benefit from the fast path (which is the only way to avoid any allocation). We really need a trait for sortable types (separate issue of course).

nalimilan commented:
@JackDevine to fix the test failures, I think you need to add this line to the top of the file:
```julia
isdefined(Main, :TestHelpers) || @eval Main include(joinpath(dirname(@__FILE__), "TestHelpers.jl"))
```

JackDevine committed:
Allow `AbstractString` and `Symbol` access to the fast path.

Also, make sure that `OffsetArray` is available inside `\test\sets.jl`.

JackDevine committed:
Allow `AbstractString` and `Symbol` access to the fast path.

Also, make sure that `OffsetArray` is available inside `\test\sets.jl`.

JackDevine commented:
@nalimilan , thanks for that code, the build is now working on my machine, so hopefully it will work on CI.

I have now included `AbstractString` and `Symbol`.

> Then there are other sortable types, like Nullable wrappers. Also, it would be really unfortunate that non-Base types would not benefit from the fast path (which is the only way to avoid any allocation). We really need a trait for sortable types (separate issue of course).

I totally agree. Something that I have thought of is that actually the fast path works for grouped data, e.g.
```julia
vec = [1,1,"z","z","w","w",4,4]
```
`vec` is obviously not sorted, but you could still send it through the fast path and get the correct result. The only problem is that to check if something is grouped, you need to use a hash like we do in the slow path, which kind of defeats the whole purpose of the fast path. But maybe your traits idea could do the trick for grouped data?

JackDevine committed:
Make sure that TestHelpers gets uncluded in Main

JackDevine committed:
Make sure that TestHelpers gets uncluded in Main

nalimilan commented on test/sets.jl:
```diff
@@ -1,5 +1,9 @@
 
+# OffsetVector is used to test `unique!` on non-standard indices (PR #20619).
+isdefined(Main, :TestHelpers) || (@eval Main include(joinpath(dirname(@__FILE__),
+   "TestHelpers.jl")))
```
Line must be aligned with `dirname` since that's an argument to the same function. Though here probably better break before `(@eval`, or even better use an `if isdefined...` multiline block.

JackDevine replied:
Woops, only just read this, if I try it inside `if isdefined...`, then the module doesn't get loaded outside the scope of the `if` block.

JackDevine committed:
Added `using TestHelpers` to make sure that `TestHelpers` loads.

JackDevine committed:
Added `using TestHelpers` to make sure that `TestHelpers` loads.

nalimilan commented on base/set.jl:
````diff
@@ -190,6 +190,75 @@ function unique(f::Callable, C)
+ 5
+```
+"""
+function unique!(A::AbstractVector{<:Union{Real, AbstractString, Symbol}})
````
Strictly speaking, this should be `::Union{AbstractVector{<:Real}, AbstractVector{<:AbstractString}, AbstractVector{<:Symbol}}`, since `issorted([1, "a"])` throws an error.

JackDevine replied:
Good point, thanks.

nalimilan commented:
> The only problem is that to check if something is grouped, you need to use a hash like we do in the slow path, which kind of defeats the whole purpose of the fast path. But maybe your traits idea could do the trick for grouped data?

Yes, that's tricky, and I'm not sure it really common to have data like this. In general arrays with a heterogeneous element type are slower than those with a homogeneous element type, so it's not really a problem if the fast path does not apply to them.

JackDevine committed:
Removed `using TestHelpers` and fixed the method signature for `unique!`

JackDevine committed:
Removed `using TestHelpers` and fixed the method signature for `unique!`

JackDevine commented:
> Yes, that's tricky, and I'm not sure it really common to have data like this.

Yes, I don't think that it is very common either. I haven't seen anyone asking for `isgrouped`, `group`, `group!` and it doesn't seem worthwhile for them to share a trait. On the other hand I think that the sorted trait sounds like a really useful idea.

JackDevine committed:
Added `using TestHelpers` again after a local test failed.

JackDevine committed:
Added `using TestHelpers` again after a local test failed.

nalimilan commented:
Would be nice to add a mention in NEWS.md.

JackDevine committed:
Rebased and added to NEWS.md under v0.7.0 Library improvements.

JackDevine committed:
Merge branch 'master' of https://github.com/JackDevine/julia

JackDevine committed:
Merge branch 'master' into master

JackDevine commented:
I can squash all of those commits if you like, sorry if I just blasted your emails with all of those commits, it is just that I was going to wait for the other reviewers to approve before I did the squash.

tkelman commented on test/sets.jl:
```diff
@@ -1,5 +1,17 @@
+
+using TestHelpers
+
+# OffsetVector is used to test `unique!` on non-standard indices (PR #20619).
```
rebase mistake that this is done twice?

JackDevine replied:
Yes, sorry about that.

JackDevine committed:
Fixed a rebase mistake where I loaded a module twice.

JackDevine committed:
Updated NEW.md so that it links to this PR.

JackDevine committed:
Changed the link to the issue, not the PR.

JackDevine committed:
Removed the manual insertion of the footnote

Footnotes are automatically generated.

nalimilan commented on NEWS.md:
```diff
@@ -65,9 +65,12 @@ Library improvements
   * `logspace` now accepts a `base` keyword argument to specify the base of the logarithmic
     range. The base defaults to 10 ([#22310]).
 
+
```
Unrelated.

JackDevine replied:
Fixed

nalimilan commented on base/set.jl:
```diff
@@ -201,6 +201,76 @@ function unique(f::Callable, C)
+
+julia> A = [7, 3, 2, 3, 7, 5];
+
+julia> unique!(A);
```
Doesn't `unique!` return `A`? If not, it should, and this should be tested (using `===`).

JackDevine replied:
Yes, `unique!` will return `A`, so for example
```julia
julia> A = [1,2,1,3,2];

julia> B = unique!(A)
3-element Array{Int64,1}:
 1
 2
 3

julia> B
3-element Array{Int64,1}:
 1
 2
 3
```
I have updated the docstring so that the semicolon is no longer there, so the user can see that `A` is being returned. I have also updated the tests so that they look like
```julia
u = [1,2,2,3,5,5]
@test unique!(u) === u
@test u == [1,2,3,5]
```
Therefore, the tests now check whether or not the returned value is correct.

nalimilan commented on test/sets.jl:
```diff
@@ -1,5 +1,10 @@
 # This file is a part of Julia. License is MIT: https://julialang.org/license
 
+# OffsetVector is used to test `unique!` on non-standard indices (PR #20619).
```
I'm not sure we really care about where it's used. If another tests starts using `OffsetArray`, this comment will be kind of misleading. Also we don't put references to PRs in the code, the git history is enough.

On the line below, you can also break after `||`, which might be more logical.

JackDevine replied:
Thanks, fixed.

nalimilan commented:
I wonder whether for users which do not care about the original order of the vector (and it is not already sorted), it would be faster to call `unique!(sort!(x))` than `unique!(x)`. If that's the case, it would be worth mentioning in the docstring, and maybe even adding a `sort` keyword argument for that (which would allow skipping the cost of `issorted`). Would you mind running a few benchmarks?

JackDevine committed:
Made requested changes.

1. Removed unrelated item from NEWS.md
2. Updated the docstring for `unique` and added a sort keyword that
sorts the data before calculating `unique!`
3. Updated the tests

fredrikekre commented on NEWS.md:
```diff
@@ -65,8 +65,7 @@ Library improvements
     git repo. Additionally, the argument order was changed to be consistent with the git
     command line tool ([#22062]).
 
-  * `logspace` now accepts a `base` keyword argument to specify the base of the logarithmic
```
?

JackDevine replied:
@nalimilan said that that was unrelated, I can put it back if you like.

fredrikekre replied:
I think he meant just the extra added line?

JackDevine replied:
Woops, he meant the blank line below.

JackDevine commented:
> Would you mind running a few benchmarks?

```julia
srand(1)
x = [rand(1:10_000) for _ in 1:1_000_000]
@show issorted(x)

println("Sorting the vector in place")
@btime unique!(x, sortdata=true)

srand(1)
x = [rand(1:10_000) for _ in 1:1_000_000]
println("Running the not sorted vector through the slow path")
@btime unique!(x)
```

```
issorted(x) = false
Sorting the vector in place
  79.683 μs (1 allocation: 96 bytes)
Running the not sorted vector through the slow path
  518.621 μs (23 allocations: 193.98 KiB)
```
So adding the `sortdata` keyword could be really useful for some users.

JackDevine committed:
Put the logspace news item back in.

fredrikekre commented on NEWS.md:
```diff
@@ -66,7 +66,9 @@ Library improvements
 
   * `logspace` now accepts a `base` keyword argument to specify the base of the logarithmic
-    range. The base defaults to 10 ([#22310]).
+   range. The base defaults to 10 ([#22310]).
```
one missing space before `range.`

JackDevine replied:
Good point, thanks.

JackDevine committed:
Added a space before range.

nalimilan commented:
Thanks for the benchmarks, that confirms that sorting is a really good strategy where possible.

Though I'm not sure whether we should add a `sort` keyword argument or whether telling users to call `sort!` when possible in the docstring would be enough. After all, the cost of calling `issorted` should be small compared with the cost of sorting, so calling `sort!` beforehand should be almost as efficient. If needed, the `issorted` check could be skipped in the future using a `SortedVector` wrapper (cf. the [Search & Find Julep](https://github.com/JuliaLang/Juleps/blob/master/Find.md#particular-cases)). Opinions?

tkelman commented on base/set.jl:
```diff
@@ -201,6 +201,104 @@ function unique(f::Callable, C)
+
+Remove duplicate items as determined by [`isequal`](@ref), then return the modified `A`.
+By default, `unique!` will return the elements of `A` in the order that they occur. If
+you want `unique!` to return the data sorted, then set `sort=true`. With `sort=true`,
```
`sort` or `sortdata` ?

nalimilan replied:
`sort` sounds better, `data` doesn't add anything.

JackDevine replied:
Fixed.

JackDevine committed:
Changed `sortdata` to `sort`. (nalimilan, tkelman)

Some other things also addressed:
1. Tidied up some comments and the docstring.
2. Changed the way that we use dispatch.
3. Made sure that `unique!(Float64[]) == Float64[]` and added a test for
this case (the previous implementation did not work in this case).

JackDevine committed:
Revert "Changed `sortdata` to `sort`. (nalimilan, tkelman)"

This reverts commit 8bceb2373b3305a575c64314c39bae490af749f3.

JackDevine committed:
Revert "Revert "Changed `sortdata` to `sort`. (nalimilan, tkelman)""

This reverts commit 64a0613436e54582d37419e3bbdc02c45729d1d0.

JackDevine committed:
Make sure that we return _unique!(A)

JackDevine commented:
Just with regards to the `sort` keyword, I know that the devs' opinions are more important than mine, but I thought that I would discuss some things that could possibly trip up people who are slightly more naive.

In our dispatch technique, we essentially have two methods.  One is for data that can be sorted and one is for general data. Currently, the keyword is available to both methods and both methods have the potential to run the line
```julia
return _sortedunique!(sort!(A))
```
But in reality, this line has no business inside the generic method. The reason that I put it there is the following. If the `sort` keyword argument is not available to the generic method (which makes complete sense), then you have the potential for the following error:
```julia
julia> unique!([1,"w",:e, 5], sort=true)

MethodError: no method matching unique!(::Array{Any,1}; sort=true)
Closest candidates are:
  unique!(::Any) at C:\Users\devja964\Documents\Dev\julia\base\set.jl:298 got unsupported keyword argument "sort"
  unique!(!Matched::Union{AbstractArray{#s7,1} where #s7<:Real, AbstractArray{#s8,1} where #s8<:AbstractString, AbstractArray{#s9,1} where #s9<:Symbol}; sort) at C:\Users\devja964\Documents\Dev\julia\base\set.jl:285
include_string(::String, ::String) at loading.jl:515
```
But with the way that I have things right now (where `sort` is available to both methods)
```julia
unique!([1,"w",:e, 5], sort=true)
MethodError: no method matching isless(::String, ::Int64)
Closest candidates are:
  isless(!Matched::AbstractFloat, ::Real) at operators.jl:98
  isless(!Matched::Real, ::Real) at operators.jl:266
  isless(::AbstractString, !Matched::AbstractString) at strings/basic.jl:124
sort!(::Array{Any,1}, ::Int64, ::Int64, ::Base.Sort.InsertionSortAlg, ::Base.Order.ForwardOrdering) at sort.jl:252
sort!(::Array{Any,1}, ::Int64, ::Int64, ::Base.Sort.MergeSortAlg, ::Base.Order.ForwardOrdering, ::Array{Any,1}) at sort.jl:341
sort!(::Array{Any,1}, ::Base.Sort.MergeSortAlg, ::Base.Order.ForwardOrdering) at sort.jl:435
#sort!#7(::Base.Sort.MergeSortAlg, ::Function, ::Function, ::Bool, ::Base.Order.ForwardOrdering, ::Function, ::Array{Any,1}) at sort.jl:494
#unique!#3(::Bool, ::Function, ::Array{Any,1}) at set.jl:299
(::#kw##unique!)(::Array{Any,1}, ::#unique!, ::Array{Any,1}) at <missing>:0
include_string(::String, ::String) at loading.jl:515
```
If a user sees the first error, then they might think that we didn't create a method for their code without realizing that their code is flawed. However, the second error is much more clear and points directly to the problem at hand.

This whole situation seems a little awkward to me, so I am in favor of not having the `sort` keyword and just ask the user to do `unique!(sort!(x))` if they don't mind having the data sorted on return.

nalimilan commented:
I agree with your reasoning, though I don't think it's a problem that `sort!` is called from inside the generic method; anyway all of this is internal. But I don't have a strong opinion regarding the `sort` keyword argument, which is why I hoped others could comment.

JackDevine commented:
Ok cool, good to see that we are on the same page 😃 

Lets wait for the others to comment.

rfourquet commented on base/set.jl:
```diff
@@ -201,6 +201,109 @@ function unique(f::Callable, C)
+# If A is not sorted, then we will need to keep track of all of the elements that we have
+# seen so far.
+function _unique!(A)
+    seen = Set{eltype(A)}()
```
`seen = similar(A)`?

rfourquet commented on base/set.jl:
````diff
@@ -201,6 +201,109 @@ function unique(f::Callable, C)
+```
+"""
+function unique!(A::Union{AbstractVector{<:Real}, AbstractVector{<:AbstractString},
+                          AbstractVector{<:Symbol}}; sort=false)
````
The following would be equivalent no?
```
function unique!(A::AbstractVector{<:Union{Real,AbstractString,Symbol}}; sort=false)`

rfourquet replied:
Oh, by going through the "outdated" comments, I see that @nalimilan objected that it is like this "since `issorted([1, "a"])` throws an error", but I don't understand, as the type of `[1, "a"]` in not a `AbstractVector{<:Union{Real,AbstractString,Symbol}}`

martinholters replied:
Not quite: E.g. a `Vector{Union{String,Symbol}}` would match the proposed signature, but not the current one.

martinholters replied:
```julia
julia> [1, "a"] isa AbstractVector{<:Union{Real,AbstractString,Symbol}}
false

julia> Union{Int,String}[1, "a"] isa AbstractVector{<:Union{Real,AbstractString,Symbol}}
true
```

rfourquet replied:
Ah thanks I understand now! Wouldn't it be useful then to have a construct similar to `Union`, say `TypeList`, which represents exactly one of the types in the list (not the union of a subset), so that `AbstractVector{<:TypeList{Real,AbstractString,Symbol}}` would do what is needed here?

nalimilan replied:
Maybe, but I'm not sure how often this comes up. This is kind of a hack anyway due to the lack of a `Sortable` trait.

rfourquet commented on base/set.jl:
```diff
@@ -201,6 +201,109 @@ function unique(f::Callable, C)
+# elements of A with the unique elements that we see in the order that we see them. Once
+# we have iterated through A, we resize A based on the number of unique elements that we
+# see.
+function _sortedunique!(A)
```
Even if it is obvious from the context, I think it would help readers to make some assumptions explicit, e.g. `A::AbstractArray`, and that a precondition is that `!isempty(A)`.

nalimilan replied:
+1 for `::AbstractArray`, but I'm not sure about `isempty`: the code should work on empty arrays too, even if it's a waste of time, shouldn't it?

rfourquet replied:
for example `first(A)` doesn't work if `isempty(A)`. The empty case is handled in `unique!`

rfourquet replied:
I guess a real check with a call to `isempty` is negligible, but even a comment like `# precondition: !isempty(y)` would be fine

nalimilan replied:
Right, maybe better add a redundant check, which is clearer and more robust.

JackDevine replied:
I added the annotation to both `_unique!` and `_groupedunique!`, are you sure that `AbstractVector` wouldn't be better?
Also, added the redundant check.

nalimilan replied:
Yes, sorry, `AbstractVector` would be better as you cannot resize matrices.

rfourquet commented on base/set.jl:
```diff
@@ -201,6 +201,109 @@ function unique(f::Callable, C)
+    y = first(A)
+    state = start(idxs)
+    j, state = next(idxs, state)
+    for i in idxs
```
Or otherwise, why not use `A[i]` everywhere? (would it be the cost of bound checking?)

nalimilan replied:
Makes sense to do `for x in A`, and rename `j` to `i` like in the other function.

JackDevine replied:
Thanks, fixed.

rfourquet commented on test/sets.jl:
```diff
@@ -1,5 +1,9 @@
 # This file is a part of Julia. License is MIT: https://julialang.org/license
 
+isdefined(Main, :TestHelpers) ||
+                        @eval Main include(joinpath(dirname(@__FILE__), "TestHelpers.jl"))
```
The indentation of `@eval` seems strange here.

JackDevine replied:
This was bugging me too, I have now stolen the style from [arrayops.jl](https://github.com/JuliaLang/julia/blob/master/test/arrayops.jl)

rfourquet commented:
I don't have a strong opinion on the `sort` keyword, but I think I prefer the following alternative. With your code, if my input is an already sorted vector, the call to `issorted` will be significant, but passing `sort=true` will be even more costly (at least with quicksort I think), so in this case the current design doesn't offer a satisfying solution. What about an `sorted=true` keyword to indicate that the data is sorted? When the user wishes to have its data sorted, it will be more verbose than with your solution: `unique!(sort!(A), sorted=true)`, but this allows also an optimal path to the case when `A` is already sorted. The problem is that the default would have to be `sorted=false`, (false meaning "unknown"), so there is no way for the user to tell that the data is not sorted (and avoid calling the `issorted` from within `unique!`); ideally we would need a third boolean value, `unknown`.

nalimilan commented:
> What about an sorted=true keyword to indicate that the data is sorted? When the user wishes to have its data sorted, it will be more verbose than with your solution: unique!(sort!(A), sorted=true), but this allows also an optimal path to the case when A is already sorted. The problem is that the default would have to be sorted=false, (false meaning "unknown"), so there is no way for the user to tell that the data is not sorted (and avoid calling the issorted from within unique!); ideally we would need a third boolean value, unknown.

As I noted [above](https://github.com/JuliaLang/julia/pull/20619#issuecomment-308037583), I think we need a generic way to declare that a vector is sorted, via a `SortedVector` type. That's why I'd rather not add a keyword argument here for now. But it's certainly safe to drop the `sort` keyword argument for now, it will be easier to add than to remove later.

rfourquet commented:
I realize that an array doesn't need to be sorted to benefit from the fast path, it's enough that duplicates are "grouped" (cf. a previous comment), so I would consider useful to have an `grouped=true` keyword argument, as there will probably never be a `GroupedVector` abstraction. And if no keyword is passed, it seems fine to only check for `issorted`. It could for sure be added at a later stage, but now is also as good a time as any to discuss it ;-)

nalimilan commented:
Maybe, but as we discussed above, how common is it to have unsorted "grouped" vectors? Unless we have a concrete example in mind, I'd rather not introduce a keyword argument (which can easily be done later if needed).

JackDevine committed:
Made suggestions (nalimilan, rfourqet)

1. Added the annotation `::AbstractArray` to `_unique!` and
`_groupedunique!`.

2. Changed `for i in idxs` to `for x in A`

3. Fixed up the code that includes `TestHelpers`

JackDevine commented:
> I don't know how frequent it is, but I don't see why we should limit the applicability of the fast path to sorted vectors.
I don't mind waiting until someone needs it (and knows about the issue), but the time to check if a sorted vector is sorted is not negligible compared to sorting the vector. So I would favor introducing the keyword now (or a uniquegrouped! function ;-)

I have changed `_sortedunique!` to `_groupedunique!`, so getting a `groupedunique!` function would just mean removing an underscore, adding a docstring and adding to exports. `groupedunique!` would behave sort of like the C++ function.

Maybe we should wait for someone with a strong opinion on the `sort` keyword before removing it. Although, FWIW, the more I think about it the less that I am in favor of the keyword.

nalimilan commented on base/set.jl:
```diff
@@ -201,6 +201,109 @@ function unique(f::Callable, C)
+ 2
+ 5
+
+julia> A
```
I find it a bit too verbose to print the contents of `A` again. People should be familiar with in-place methods ending with `!`, no need to show this in every docstring. If in doubt, they can easily check that `A` is modified. Same for `B`.

nalimilan commented:
> Maybe we should wait for someone with a strong opinion on the sort keyword before removing it. Although, FWIW, the more I think about it the less that I am in favor of the keyword.

I'd say remove it in the interest of merging this PR, and we can discuss this elsewhere if somebody really finds it useful. Though please mention in the docstring that calling `sort!` beforehand will be more efficient (when it's possible).

nalimilan commented:
@rfourquet That idea is interesting. But I'd rather merge this as-is, and discuss in another PR possible improvements.

JackDevine committed:
Made suggestions (nalimilan)

1. Removed the `sort` keyword argument.
2. Changed `AbstractArray` to `AbstractVector`.

fredrikekre commented:
Thanks @JackDevine for working on this for almost 4 months with all the changes back and forth.

StefanKarpinski commented:
Yes, 💯 % thank you so much for sticking this out – this is a great addition that I've personally often wanted.

JackDevine commented:
Hi everyone, thanks for all of your kind suggestions and feedback. It's a great community and I've enjoyed working with you and being mentored by the Julians themselves.
